File: apps/api/src/env.ts
/**
 * Configuration d'exécution, lue une seule fois au démarrage et validée
 * strictement : une variable manquante en production arrête le processus plutôt
 * que de laisser tourner un service à moitié configuré.
 *
 * Les problèmes sont TOUS rassemblés avant d'arrêter, et non levés au premier
 * rencontré. Sur un hébergeur, chaque démarrage raté coûte un déploiement et
 * une lecture de journal : apprendre les variables manquantes une par une fait
 * autant d'allers-retours qu'il en manque.
 */

type Mode = "development" | "test" | "production";

/*
 * Sur un hébergeur, l'absence de NODE_ENV ne doit surtout pas valoir
 * « développement ». Heroku définit `DYNO` sur chaque dyno : s'y fier écarte le
 * pire des cas — un service qui démarre en ligne avec les réglages du
 * développement, se rabat silencieusement sur SQLite, puis échoue sur un module
 * absent du slug au lieu de dire ce qui manque. Le diagnostic coûtait alors un
 * déploiement et une lecture de journal pour une variable oubliée.
 */
const surUnHebergeur = (process.env.DYNO ?? "") !== "";
const mode = (process.env.NODE_ENV ?? (surUnHebergeur ? "production" : "development")) as Mode;

/** Ce qui empêche de démarrer, accumulé puis rapporté d'un bloc. */
const problemes: { variable: string; raison: string; remede: string }[] = [];

function required(name: string, fallbackInDev?: string): string {
  const value = process.env[name];
  if (value !== undefined && value !== "") return value;
  if (fallbackInDev !== undefined && mode !== "production") return fallbackInDev;
  problemes.push({
    variable: name,
    raison: "absente",
    remede: REMEDES[name] ?? "à définir dans la configuration de l'application",
  });
  return "";
}

/** Comment obtenir chaque valeur, dit une fois plutôt que cherché dans un guide. */
const REMEDES: Readonly<Record<string, string>> = {
  JWT_SECRET: "une chaîne aléatoire d'au moins 32 caractères — `openssl rand -base64 32`",
  MEDIA_SIGNING_SECRET: "une autre chaîne aléatoire — `openssl rand -base64 32`",
  DATABASE_URL: "fournie par l'add-on Heroku Postgres, à attacher à l'application",
};

function optional(name: string): string | undefined {
  const value = process.env[name];
  return value === undefined || value === "" ? undefined : value;
}

function int(name: string, fallback: number): number {
  const raw = process.env[name];
  if (raw === undefined || raw === "") return fallback;
  const parsed = Number.parseInt(raw, 10);
  if (Number.isNaN(parsed)) throw new Error(`${name} doit être un entier`);
  return parsed;
}

/**
 * Résout un chemin SQLite relatif par rapport au paquet `apps/api`, et non par
 * rapport au répertoire courant : les tests et les scripts peuvent ainsi être
 * lancés depuis la racine du dépôt comme depuis le paquet lui-même.
 */
function resolveSqliteUrl(raw: string): string {
  if (!raw.startsWith("file:")) return raw;
  const path = raw.slice("file:".length);
  if (path.startsWith("/") || path === ":memory:") return raw;
  const packageRoot = new URL("..", import.meta.url).pathname.replace(/\/$/, "");
  return `file:${packageRoot}/${path.replace(/^\.\//, "")}`;
}

const driver = (
  process.env.WEAVE_DB ?? (mode === "production" ? "postgres" : "sqlite")
).toLowerCase();

if (driver !== "postgres" && driver !== "sqlite") {
  throw new Error(`WEAVE_DB doit valoir "postgres" ou "sqlite" (reçu : ${driver})`);
}
if (driver === "sqlite" && mode === "production") {
  problemes.push({
    variable: "WEAVE_DB",
    raison: "vaut « sqlite », réservé au développement",
    remede: "définir WEAVE_DB=postgres",
  });
}

export const env = {
  mode,
  isProduction: mode === "production",
  port: int("PORT", 3000),
  webOrigin: process.env.PUBLIC_WEB_ORIGIN ?? "http://localhost:5173",
  /**
   * Répertoire du site vitrine compilé. Quand il est présent, l'API le sert
   * elle-même : un seul processus, donc un seul dyno à payer. En développement
   * le site est servi par Vite et cette variable reste vide.
   */
  webDist: optional("WEB_DIST_PATH"),

  db: {
    driver: driver as "postgres" | "sqlite",
    postgresUrl: driver === "postgres" ? required("DATABASE_URL") : optional("DATABASE_URL"),
    sqliteUrl: resolveSqliteUrl(process.env.DATABASE_URL_SQLITE ?? "file:./prisma/dev.db"),
    /**
     * Même compromis que pour Redis : Heroku Postgres présente un certificat
     * auto-signé. La connexion reste chiffrée, l'identité du serveur n'est pas
     * vérifiée. Réservé aux bases jointes par le réseau interne de l'hébergeur.
     */
    sslInsecure: process.env.DATABASE_SSL_INSECURE === "true",
  },

  redis: {
    // Heroku Key-Value Store expose deux adresses : `REDIS_TLS_URL` (chiffrée,
    // obligatoire sur les petits plans) et `REDIS_URL`. On préfère la première
    // quand elle existe.
    url: process.env.REDIS_TLS_URL ?? process.env.REDIS_URL ?? "redis://localhost:6379",
    /**
     * Accepte un certificat serveur non vérifiable.
     *
     * Nécessaire chez Heroku, dont le magasin clé-valeur présente un certificat
     * auto-signé : sans cela la connexion échoue. Le trafic reste chiffré, mais
     * l'identité du serveur n'est pas vérifiée — à n'activer que sur un réseau
     * interne d'hébergeur, jamais pour joindre un cache à travers l'internet
     * public.
     */
    tlsInsecure: process.env.REDIS_TLS_INSECURE === "true",
  },

  auth: {
    jwtSecret: required("JWT_SECRET", "secret-de-developpement-non-utilisable-en-production"),
    accessTtlSeconds: int("ACCESS_TOKEN_TTL_SECONDS", 900),
    refreshTtlDays: int("REFRESH_TOKEN_TTL_DAYS", 60),
  },

  media: {
    baseUrl: process.env.MEDIA_BASE_URL ?? "http://localhost:3000/media",
    signingSecret: required("MEDIA_SIGNING_SECRET", "secret-media-developpement"),
    signedUrlTtlSeconds: int("MEDIA_URL_TTL_SECONDS", 600),
  },

  apns: {
    keyId: optional("APNS_KEY_ID"),
    teamId: optional("APNS_TEAM_ID"),
    bundleId: process.env.APNS_BUNDLE_ID ?? "com.weave.app",
    keyPath: optional("APNS_KEY_PATH"),
    environment: (process.env.APNS_ENVIRONMENT ?? "sandbox") as "sandbox" | "production",
    get configured(): boolean {
      return Boolean(
        process.env.APNS_KEY_ID && process.env.APNS_TEAM_ID && process.env.APNS_KEY_PATH,
      );
    },
  },

  appStore: {
    issuerId: optional("APPSTORE_ISSUER_ID"),
    keyId: optional("APPSTORE_KEY_ID"),
    keyPath: optional("APPSTORE_KEY_PATH"),
    environment: (process.env.APPSTORE_ENVIRONMENT ?? "sandbox") as "sandbox" | "production",
    get configured(): boolean {
      return Boolean(
        process.env.APPSTORE_ISSUER_ID &&
        process.env.APPSTORE_KEY_ID &&
        process.env.APPSTORE_KEY_PATH,
      );
    },
  },
} as const;

if (env.isProduction && env.auth.jwtSecret !== "" && env.auth.jwtSecret.length < 32) {
  problemes.push({
    variable: "JWT_SECRET",
    raison: `trop court (${env.auth.jwtSecret.length} caractères, minimum 32)`,
    remede: REMEDES.JWT_SECRET!,
  });
}

/**
 * Rapport de démarrage.
 *
 * Écrit sur la sortie d'erreur, seule voie que l'on soit sûr de retrouver dans
 * le journal d'un hébergeur, et sous une forme qui se lit sur un téléphone :
 * une ligne par variable, le remède à côté du problème.
 */
if (problemes.length > 0) {
  const lignes = [
    "",
    "  Weave ne peut pas démarrer : la configuration est incomplète.",
    "",
    ...problemes.flatMap((probleme) => [
      `  • ${probleme.variable} — ${probleme.raison}`,
      `      ${probleme.remede}`,
    ]),
    "",
    "  Sur Heroku : tableau de bord → Settings → Config Vars,",
    "  ou le workflow « Heroku — configurer les variables » depuis GitHub.",
    "",
  ];
  console.error(lignes.join("\n"));
  process.exit(1);
}

/**
 * Avertissements : ce qui n'empêche pas de démarrer, mais se verra aussitôt.
 *
 * Ils ne sont émis que sur un dyno Heroku (`DYNO` est posé par la plateforme),
 * parce que ce sont ses particularités qu'ils décrivent. Ailleurs, exiger ces
 * réglages serait faux — et dangereux : accepter un certificat non vérifié n'a
 * de sens que sur un réseau interne d'hébergeur.
 */
if (process.env.DYNO !== undefined) {
  const avertissements: { sujet: string; raison: string; remede: string }[] = [];

  if (!env.db.sslInsecure) {
    avertissements.push({
      sujet: "DATABASE_SSL_INSECURE",
      raison: "la base d'Heroku présente un certificat auto-signé sur son réseau interne",
      remede: "définir DATABASE_SSL_INSECURE=true, sans quoi la base restera injoignable",
    });
  }

  // Uniquement si l'adresse est chiffrée : sur une adresse en clair, la
  // question ne se pose pas.
  if (env.redis.url.startsWith("rediss://") && !env.redis.tlsInsecure) {
    avertissements.push({
      sujet: "REDIS_TLS_INSECURE",
      raison: "le magasin clé-valeur d'Heroku présente lui aussi un certificat auto-signé",
      remede: "définir REDIS_TLS_INSECURE=true, sans quoi le cache restera injoignable",
    });
  }

  if (env.webOrigin.includes("localhost")) {
    avertissements.push({
      sujet: "PUBLIC_WEB_ORIGIN",
      raison: "pointe encore sur localhost",
      remede:
        "définir l'adresse publique du site, sinon les appels depuis le navigateur seront refusés",
    });
  }

  if (env.media.baseUrl.includes("localhost")) {
    avertissements.push({
      sujet: "MEDIA_BASE_URL",
      raison: "pointe encore sur localhost",
      remede: "définir l'adresse publique des médias, sinon les photos ne s'afficheront pas",
    });
  }

  if (avertissements.length > 0) {
    console.warn(
      [
        "",
        "  Weave démarre, mais la configuration est incomplète :",
        "",
        ...avertissements.flatMap((a) => [`  • ${a.sujet} — ${a.raison}`, `      ${a.remede}`]),
        "",
        "  Le workflow « Heroku — configurer les variables » pose tout cela.",
        "",
      ].join("\n"),
    );
  }
}

File: apps/api-rs/src/env.rs
//! Configuration d'exécution, lue une seule fois au démarrage et validée
//! strictement : une variable manquante en production arrête le processus
//! plutôt que de laisser tourner un service à moitié configuré.
//!
//! Les problèmes sont TOUS rassemblés avant d'arrêter, et non levés au premier
//! rencontré. Sur un hébergeur, chaque démarrage raté coûte un déploiement et
//! une lecture de journal : apprendre les variables manquantes une par une
//! fait autant d'allers-retours qu'il en manque.

use std::env;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Mode {
    Development,
    Test,
    Production,
}

impl Mode {
    pub fn is_production(self) -> bool {
        self == Mode::Production
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Driver {
    Postgres,
    Sqlite,
}

impl Driver {
    pub fn as_str(self) -> &'static str {
        match self {
            Driver::Postgres => "postgres",
            Driver::Sqlite => "sqlite",
        }
    }
}

/// Ce qui empêche de démarrer, accumulé puis rapporté d'un bloc.
struct Probleme {
    variable: &'static str,
    raison: String,
    remede: &'static str,
}

/// Comment obtenir chaque valeur, dit une fois plutôt que cherché dans un guide.
fn remede(nom: &str) -> &'static str {
    match nom {
        "JWT_SECRET" => "une chaîne aléatoire d'au moins 32 caractères — `openssl rand -base64 32`",
        "MEDIA_SIGNING_SECRET" => "une autre chaîne aléatoire — `openssl rand -base64 32`",
        "DATABASE_URL" => "fournie par l'add-on Heroku Postgres, à attacher dans Resources",
        "REDIS_URL" => "fournie par l'add-on Heroku Key-Value Store, à attacher dans Resources",
        _ => "à définir dans la configuration de l'application",
    }
}

/// Lit une variable obligatoire, ou note ce qui manque pour le rapport final.
fn exiger(
    nom: &'static str,
    defaut_dev: Option<&str>,
    mode: Mode,
    problemes: &mut Vec<Probleme>,
) -> String {
    if let Some(v) = lire(nom) {
        return v;
    }
    if let Some(d) = defaut_dev {
        if mode != Mode::Production {
            return d.to_string();
        }
    }
    problemes.push(Probleme {
        variable: nom,
        raison: "absente".to_string(),
        remede: remede(nom),
    });
    String::new()
}

fn entier(nom: &str, defaut: i64) -> i64 {
    lire(nom).and_then(|v| v.parse().ok()).unwrap_or(defaut)
}

fn lire(nom: &str) -> Option<String> {
    match env::var(nom) {
        Ok(v) if !v.is_empty() => Some(v),
        _ => None,
    }
}

#[derive(Debug, Clone)]
pub struct Db {
    pub driver: Driver,
    pub url: String,
    /// Heroku présente un certificat auto-signé sur son réseau interne : la
    /// connexion reste chiffrée, l'identité du serveur n'est pas vérifiée.
    pub ssl_insecure: bool,
}

#[derive(Debug, Clone)]
pub struct Cache {
    pub url: String,
    pub tls_insecure: bool,
}

#[derive(Debug, Clone)]
pub struct Media {
    pub signing_secret: String,
    pub base_url: String,
    /// Les URL de médias sont courtes par conception : une adresse qui fuite
    /// ne doit pas rester lisible longtemps.
    pub ttl_url_signee_secondes: i64,
}

#[derive(Debug, Clone)]
pub struct Auth {
    pub jwt_secret: String,
    /// Court : l'accès expire vite, le rafraîchissement prend le relais.
    pub access_ttl_secondes: i64,
    pub refresh_ttl_jours: i64,
}

#[derive(Debug, Clone)]
pub struct Apns {
    pub key_id: Option<String>,
    pub team_id: Option<String>,
    pub key_path: Option<String>,
    pub bundle_id: String,
    pub environnement: String,
    /// Sans ces trois valeurs, aucune notification ne peut être signée : les
    /// envois sont alors simulés plutôt que de faire échouer les requêtes.
    pub configure: bool,
}

#[derive(Debug, Clone)]
pub struct AppStore {
    pub environnement: String,
    /// Sans ces trois valeurs, aucune transaction ne peut être vérifiée
    /// auprès d'Apple — et en production, aucune ne doit être acceptée.
    pub configure: bool,
}

#[derive(Debug, Clone)]
pub struct Env {
    pub mode: Mode,
    pub port: u16,
    pub db: Db,
    pub cache: Cache,
    pub media: Media,
    pub auth: Auth,
    pub apns: Apns,
    pub app_store: AppStore,
    pub web_origin: String,
    pub web_dist: Option<String>,
}

impl Env {
    pub fn is_production(&self) -> bool {
        self.mode.is_production()
    }
}

/// Lit la configuration, ou décrit d'un bloc tout ce qui l'empêche.
pub fn charger() -> Result<Env, String> {
    // Sur un hébergeur, l'absence de NODE_ENV ne doit surtout pas valoir
    // « développement ». Heroku définit `DYNO` sur chaque dyno : s'y fier
    // écarte le pire des cas — un service qui démarre en ligne avec les
    // réglages du développement.
    let sur_un_hebergeur = lire("DYNO").is_some();
    let mode = match lire("NODE_ENV").as_deref() {
        Some("production") => Mode::Production,
        Some("test") => Mode::Test,
        Some(_) => Mode::Development,
        None if sur_un_hebergeur => Mode::Production,
        None => Mode::Development,
    };

    let mut problemes: Vec<Probleme> = Vec::new();
    let mut avertissements: Vec<Probleme> = Vec::new();

    let jwt_secret = exiger("JWT_SECRET", Some("secret-de-developpement-non-utilisable"), mode, &mut problemes);
    let signing_secret = exiger("MEDIA_SIGNING_SECRET", Some("secret-media-developpement"), mode, &mut problemes);

    let weave_db = lire("WEAVE_DB").map(|v| v.to_lowercase());
    let driver = match weave_db.as_deref() {
        Some("postgres") => Driver::Postgres,
        Some("sqlite") => Driver::Sqlite,
        Some(autre) => {
            return Err(format!(
                "WEAVE_DB doit valoir \"postgres\" ou \"sqlite\" (reçu : {autre})"
            ));
        }
        None if mode.is_production() => Driver::Postgres,
        None => Driver::Sqlite,
    };

    if driver == Driver::Sqlite && mode.is_production() {
        problemes.push(Probleme {
            variable: "WEAVE_DB",
            raison: "vaut « sqlite », réservé au développement".to_string(),
            remede: "définir WEAVE_DB=postgres",
        });
    }

    let db_url = match driver {
        Driver::Postgres => exiger("DATABASE_URL", None, mode, &mut problemes),
        Driver::Sqlite => lire("DATABASE_URL_SQLITE")
            .unwrap_or_else(|| "sqlite://./weave-dev.sqlite?mode=rwc".to_string()),
    };

    let redis_url = exiger("REDIS_URL", Some("redis://127.0.0.1:6379"), mode, &mut problemes);

    let ssl_insecure = lire("DATABASE_SSL_INSECURE").is_some_and(|v| v == "true");
    let tls_insecure = lire("REDIS_TLS_INSECURE").is_some_and(|v| v == "true");

    // Chez Heroku, ces deux réglages ne sont pas un confort : sans eux la base
    // et le cache restent injoignables derrière leurs certificats auto-signés.
    if mode.is_production() && sur_un_hebergeur {
        if !ssl_insecure {
            avertissements.push(Probleme {
                variable: "DATABASE_SSL_INSECURE",
                raison: "la base d'Heroku présente un certificat auto-signé sur son réseau interne"
                    .to_string(),
                remede: "définir DATABASE_SSL_INSECURE=true, sans quoi la base restera injoignable",
            });
        }
        if !tls_insecure {
            avertissements.push(Probleme {
                variable: "REDIS_TLS_INSECURE",
                raison: "le magasin clé-valeur d'Heroku présente lui aussi un certificat auto-signé"
                    .to_string(),
                remede: "définir REDIS_TLS_INSECURE=true, sans quoi le cache restera injoignable",
            });
        }
    }

    let web_origin = lire("PUBLIC_WEB_ORIGIN").unwrap_or_else(|| "http://localhost:5173".to_string());
    let media_base = lire("MEDIA_BASE_URL").unwrap_or_else(|| "http://localhost:3000/media".to_string());

    if mode.is_production() {
        if web_origin.contains("localhost") {
            avertissements.push(Probleme {
                variable: "PUBLIC_WEB_ORIGIN",
                raison: "pointe encore sur localhost".to_string(),
                remede: "définir l'adresse publique du site, sinon les appels depuis le navigateur seront refusés",
            });
        }
        if media_base.contains("localhost") {
            avertissements.push(Probleme {
                variable: "MEDIA_BASE_URL",
                raison: "pointe encore sur localhost".to_string(),
                remede: "définir l'adresse publique des médias, sinon les photos ne s'afficheront pas",
            });
        }
    }

    if !problemes.is_empty() {
        return Err(rapport(
            "Weave ne peut pas démarrer : la configuration est incomplète.",
            &problemes,
        ));
    }

    if !avertissements.is_empty() {
        eprint!(
            "{}",
            rapport(
                "Weave démarre, mais la configuration est incomplète :",
                &avertissements
            )
        );
    }

    let port = lire("PORT")
        .and_then(|p| p.parse().ok())
        .unwrap_or(3000_u16);

    Ok(Env {
        mode,
        port,
        db: Db {
            driver,
            url: db_url,
            ssl_insecure,
        },
        cache: Cache {
            url: redis_url,
            tls_insecure,
        },
        media: Media {
            signing_secret,
            base_url: media_base,
            ttl_url_signee_secondes: entier("MEDIA_URL_TTL_SECONDS", 600),
        },
        apns: Apns {
            key_id: lire("APNS_KEY_ID"),
            team_id: lire("APNS_TEAM_ID"),
            key_path: lire("APNS_KEY_PATH"),
            bundle_id: lire("APNS_BUNDLE_ID").unwrap_or_else(|| "com.weave.app".to_string()),
            environnement: lire("APNS_ENVIRONMENT").unwrap_or_else(|| "sandbox".to_string()),
            configure: lire("APNS_KEY_ID").is_some()
                && lire("APNS_TEAM_ID").is_some()
                && lire("APNS_KEY_PATH").is_some(),
        },
        app_store: AppStore {
            environnement: lire("APPSTORE_ENVIRONMENT").unwrap_or_else(|| "sandbox".to_string()),
            configure: lire("APPSTORE_ISSUER_ID").is_some()
                && lire("APPSTORE_KEY_ID").is_some()
                && lire("APPSTORE_KEY_PATH").is_some(),
        },
        auth: Auth {
            jwt_secret,
            access_ttl_secondes: entier("ACCESS_TOKEN_TTL_SECONDS", 900),
            refresh_ttl_jours: entier("REFRESH_TOKEN_TTL_DAYS", 60),
        },
        web_origin,
        web_dist: lire("WEB_DIST_PATH"),
    })
}

fn rapport(titre: &str, problemes: &[Probleme]) -> String {
    let mut s = format!("\n  {titre}\n\n");
    for p in problemes {
        s.push_str(&format!("  • {} — {}\n      {}\n", p.variable, p.raison, p.remede));
    }
    s.push_str("\n  Sur Heroku : tableau de bord → Settings → Config Vars,\n");
    s.push_str("  ou le workflow « Heroku — configurer les variables » depuis GitHub.\n\n");
    s
}
